diff --git a/src/common/netif.cpp b/src/common/netif.cpp
index 8c0c4fa..3762da8 100644
--- a/src/common/netif.cpp
+++ b/src/common/netif.cpp
@@ -24,7 +24,7 @@
 #endif
 #elif defined(WIN32)
 #include <iphlpapi.h>
-#elif defined(__APPLE__)
+#elif defined(__APPLE__) && __has_include(<net/route.h>)
 #include <net/route.h>
 #include <sys/sysctl.h>
 #endif
@@ -226,7 +226,7 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
     return std::nullopt;
 }
 
-#elif defined(__APPLE__)
+#elif defined(__APPLE__) && __has_include(<net/route.h>)
 
 #define ROUNDUP32(a) \
     ((a) > 0 ? (1 + (((a) - 1) | (sizeof(uint32_t) - 1))) : sizeof(uint32_t))
