diff --git a/src/common/netif.cpp b/src/common/netif.cpp
index 8c0c4fa0832..fcc6b8ffa73 100644
--- a/src/common/netif.cpp
+++ b/src/common/netif.cpp
@@ -25,9 +25,12 @@
 #elif defined(WIN32)
 #include <iphlpapi.h>
 #elif defined(__APPLE__)
+#include <TargetConditionals.h>
+#if TARGET_OS_OSX
 #include <net/route.h>
 #include <sys/sysctl.h>
 #endif
+#endif
 
 #ifdef HAVE_IFADDRS
 #include <sys/types.h>
@@ -226,7 +229,7 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
     return std::nullopt;
 }
 
-#elif defined(__APPLE__)
+#elif defined(__APPLE__) && TARGET_OS_OSX
 
 #define ROUNDUP32(a) \
     ((a) > 0 ? (1 + (((a) - 1) | (sizeof(uint32_t) - 1))) : sizeof(uint32_t))
